geode-gfsh/src/distributedTest/java/org/apache/geode/management/internal/cli/commands/RebalanceMembersColocationTest.java

```diff
@@ -0,0 +1,97 @@
+
+import junitparams.JUnitParamsRunner;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.internal.cache.PartitionAttributesImpl;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.junit.categories.PersistenceTest;
+import org.apache.geode.test.junit.rules.GfshCommandRule;
+
+
+@Category(PersistenceTest.class)
+@RunWith(JUnitParamsRunner.class)
+public class RebalanceMembersColocationTest {
+
+  public static final String PARENT_REGION_NAME = "parentRegion";
+  public static final String CHILD_REGION_NAME = "childRegion";
+  @Rule
+  public ClusterStartupRule cluster = new ClusterStartupRule();
+
+  @Rule
+  public GfshCommandRule gfsh = new GfshCommandRule();
+
+  @Test
+  public void testRebalanceResultOutputMemberCountWithColocatedRegions() throws Exception {
+    MemberVM locator = cluster.startLocatorVM(0);
+
+    MemberVM server1 = cluster.startServerVM(1, locator.getPort());
+    MemberVM server2 = cluster.startServerVM(2, locator.getPort());
+
+    server1.invoke(() -> {
+      Region<Object, Object> parentRegion = Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.PARTITION).create(PARENT_REGION_NAME);
+
+      IntStream.range(0, 500).forEach(i -> parentRegion.put("key" + i, "value" + 1));
+
+      PartitionAttributesImpl attributes = new PartitionAttributesImpl();
+      attributes.setColocatedWith(PARENT_REGION_NAME);
+
+      Region<Object, Object> childRegion = Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.PARTITION).setPartitionAttributes(attributes)
+          .create(CHILD_REGION_NAME);
+
+      IntStream.range(0, 500).forEach(i -> childRegion.put("key" + i, "value" + 1));
+    });
+
+    server2.invoke(() -> {
+      Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.PARTITION).create(PARENT_REGION_NAME);
+
+      PartitionAttributesImpl attributes = new PartitionAttributesImpl();
+      attributes.setColocatedWith(PARENT_REGION_NAME);
+
+      Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.PARTITION).setPartitionAttributes(attributes)
+          .create(CHILD_REGION_NAME);
+    });
+
+    locator.waitUntilRegionIsReadyOnExactlyThisManyServers("/" + PARENT_REGION_NAME, 2);
+    locator.waitUntilRegionIsReadyOnExactlyThisManyServers("/" + CHILD_REGION_NAME, 2);
+
+    gfsh.connectAndVerify(locator);
+
+    Map<String, List<String>> rebalanceResult = gfsh.executeAndAssertThat("rebalance")
+        .statusIsSuccess().hasTableSection().getActual().getContent();
+
+    assertThat(rebalanceResult.get("Value").get(9)).isEqualTo("2");
```

Comment:
It would be better to assert that the output for both regions is correct rather than just the first one that appears in the gfsh output. I think that the child region still reports 0 for the total members executed on, which is not correct.

Reply:
I'm not sure but I think that it is normal to get zeros for colocated region as the whole rebalance is executed on parent region. 
As I see when execute rebalance it add all results to parent region. It goes in this part of code
```
for (PartitionRebalanceInfo details : detailSet) {
   results.addDetails(details);
}

```
If it is not normal behaviour than this whole logic is wrong and should be changed.

Reply:
I think that it's incorrect for "number of members rebalance was executed on" to be zero for colocated regions, since the rebalance *is* being executed on members for those regions. The other details like buckets moved and primaries transferred aren't zero for colocated regions, so members executed on shouldn't be zero either.

Reply:
Hi @DonalEvans,
I tested this without changes due to adding "number of members rebalance was executed on" row and got the following output:
```

Command result for <rebalance>: 
Rebalanced partition regions /parentRegion

                                       Rebalanced Stats                                         | Value
----------------------------------------------------------------------------------------------- | -----
Total bytes in all redundant bucket copies created during this rebalance                        | 0
Total time (in milliseconds) spent creating redundant bucket copies during this rebalance       | 0
Total number of redundant copies created during this rebalance                                  | 0
Total bytes in buckets moved during this rebalance                                              | 18426
Total time (in milliseconds) spent moving buckets during this rebalance                         | 4119
Total number of buckets moved during this rebalance                                             | 132
Total time (in milliseconds) spent switching the primary state of buckets during this rebalance | 0
Total primaries transferred during this rebalance                                               | 0
Total time (in milliseconds) for this rebalance                                                 | 8372

Rebalanced partition regions /childRegion

                                       Rebalanced Stats                                         | Value
----------------------------------------------------------------------------------------------- | -----
Total bytes in all redundant bucket copies created during this rebalance                        | 0
Total time (in milliseconds) spent creating redundant bucket copies during this rebalance       | 0
Total number of redundant copies created during this rebalance                                  | 0
Total bytes in buckets moved during this rebalance                                              | 0
Total time (in milliseconds) spent moving buckets during this rebalance                         | 0
Total number of buckets moved during this rebalance                                             | 0
Total time (in milliseconds) spent switching the primary state of buckets during this rebalance | 0
Total primaries transferred during this rebalance                                               | 0
Total time (in milliseconds) for this rebalance                                                 | 0
```

From this it seems that all details are zero. Also what I saw is that stats for parentRegion is twice updated with stats from both regions, parentRegion and childRegion.  
Before breaking Gfsh out into separate module it print result for both regions together, as it is below.

```
Command result for <rebalance>: 
Rebalanced partition regions /parentRegion, /childRegion

                                       Rebalanced Stats                                         | Value
----------------------------------------------------------------------------------------------- | -----
Total bytes in all redundant bucket copies created during this rebalance                        | 0
Total time (in milliseconds) spent creating redundant bucket copies during this rebalance       | 0
Total number of redundant copies created during this rebalance                                  | 0
Total bytes in buckets moved during this rebalance                                              | 18426
Total time (in milliseconds) spent moving buckets during this rebalance                         | 4119
Total number of buckets moved during this rebalance                                             | 132
Total time (in milliseconds) spent switching the primary state of buckets during this rebalance | 0
Total primaries transferred during this rebalance                                               | 0
Total time (in milliseconds) for this rebalance                                                 | 8372

```
From that I can see that it is normal that output for childRegion is all zeros or the whole logic is wrong and should be changed. 
In case it is normal that result is all zeros than this change has a sense to get correct number of servers on which rebalance is executed.

Reply:
Ah, I see, thanks for the clarification. In that case, yes, the output provided for "members executed on" is consistent with the output for the other stats, and I withdraw my request for additional changes.

That being said, it might be worth discussing on the dev list whether showing 0 for colocated regions is actually correct, and filing a new bug ticket if people think that we should output differently. Having a section with all the values 0 seems potentially misleading and confusion to me.